src/plugins/auth/auth.ts

```diff
@@ -205,6 +216,48 @@ const plugin: FastifyPluginAsync<AuthPluginOptions> = async (fastify, options) =
       .catch((err) => log.warn(err, `mailer failed. link: ${link}`));
   }
 
+  async function verifyCredentials(
+    member: Member,
+    body: { email: string; password: string },
+    challenge?: string,
+  ) {
+    /* let verified is used to store the output of bcrypt.compare() 
+    bcrypt.compare() allows to compare the provided password with a stored hash. 
+    It deduces the salt from the hash and is able to then hash the provided password correctly for comparison
+    if they match, verified is true 
+    if they do not match, verified is false
+    if the user does not has a stored hash, verified is null
+    */
+    let verified;
+    if (member.password === null) {
+      verified = null;
+    } else {
+      verified = bcrypt
+        .compare(body.password, member.password)
+        .then(async (res: boolean) => {
+          if (res) {
+            // generate token with member info and expiration
+            const token = await promisifiedJwtSign({ sub: member.id, challenge }, JWT_SECRET, {
+              expiresIn: `${LOGIN_TOKEN_EXPIRATION_IN_MINUTES}m`,
+            });
+            if (challenge) {
+              // token for graasp mobile app
+              return token;
+            } else {
+              // link for graasp web
+              const linkPath = '/auth';
+              const link = `${PROTOCOL}://${EMAIL_LINKS_HOST}${linkPath}?t=${token}`;
+              return link;
```

Comment:
Maybe it makes sense to factor out the challenge check? I don't really like the fact this function either return a token, or a link. I think it can be dealt in the respective endpoint (that is, build the link if needed only).

Reply:
I refactored this. The verifyCredentials function always returns a boolean. If the password is verified then it generates the token+link for the web app and only the token for the mobile app